source/extensions/common/tap/tap_matcher.cc

```diff
@@ -121,6 +129,206 @@ void HttpHeaderMatcherBase::matchHeaders(const Http::HeaderMap& headers,
+  auto pattern_remainder = pattern.substr(pattern_index);
+  if ((0 != limit_) && (pattern_remainder.length() > (limit_ - ctx->processed_bytes_))) {
+    // Even if we got match it would be outside the search limit
+    return false;
+  }
+  return ((pattern_remainder.length() <= data.length()) && data.startsWith(pattern_remainder));
+}
+
+// Method buffers last bytes from the currently processed body in overlap_.
+// This is required to find patterns which spans across multiple body chunks.
+void HttpGenericBodyMatcher::bufferLastBytes(const Buffer::Instance& data,
+                                             HttpGenericBodyMatcherCtx* ctx) {
+  // The matcher buffers the last seen X bytes where X is equal to the length of the
+  // longest pattern - 1. With the arrival of the new 'data' the following situations
+  // are possible:
+  // 1. The new data's length is larger or equal to X. In this case just copy last X bytes
+  // from the data to overlap_ buffer.
+  // 2. The new data length is smaller than X and there is enough room in overlap buffer to just
+  // copy the bytes from data.
+  // 3. The new data length is smaller than X and there is not enough room in overlap buffer.
+  if (data.length() >= ctx->capacity_) {
+    // Case 1:
+    // Just overwrite the entire overlap_ buffer with new data.
+    ctx->overlap_.resize(ctx->capacity_);
+    data.copyOut(data.length() - ctx->capacity_, ctx->capacity_, ctx->overlap_.data());
+  } else {
+    if (data.length() <= (ctx->capacity_ - ctx->overlap_.size())) {
+      // Case 2. Just add the new data on top of already buffered.
+      const auto size = ctx->overlap_.size();
+      ctx->overlap_.resize(ctx->overlap_.size() + data.length());
+      data.copyOut(0, data.length(), ctx->overlap_.data() + size);
+    } else {
+      // Case 3. First shift data to make room for new data and then copy
+      // entire new buffer.
+      const size_t shift = ctx->overlap_.size() - (ctx->capacity_ - data.length());
+      for (size_t i = 0; i < (ctx->overlap_.size() - shift); i++) {
+        ctx->overlap_[i] = ctx->overlap_[i + shift];
+      }
+      const auto size = ctx->overlap_.size();
+      ctx->overlap_.resize(ctx->capacity_);
+      data.copyOut(0, data.length(), ctx->overlap_.data() + (size - shift));
+    }
+  }
+}
+
+// Method takes list of indexes of patterns not yet located in the http body and returns the
+// length of the longest pattern.
+// This is used by matcher to buffer as minimum bytes as possible.
+size_t HttpGenericBodyMatcher::calcLongestPatternSize(const std::list<uint32_t>& indexes) const {
+  ASSERT(!indexes.empty());
+  size_t max_len = 0;
+  for (const auto& i : indexes) {
+    max_len = std::max(max_len, patterns_->at(i).length());
+  }
+  return max_len;
+}
+
+// Method checks if it is possible to reduce the size of overlap_ buffer.
+void HttpGenericBodyMatcher::resizeOverlapBuffer(HttpGenericBodyMatcherCtx* ctx) {
```

Comment:
Can we also add some tests where one match is a subset or overlap of another?

Alternately it might be cool to have a test which randomly selects 2 strings between 1-full_string
and 2 valid offsets between 0, length_string-length_match and asserts we find both.  Given the complexity of the code I think it'd be great to have a randomized subset test which could just use your fixed body_parts_ and select those random segments.

Reply:
Done as suggested. Test generates random number of up to 10 substrings of a pattern located on the boundary of two body chunks.